crates/storage/provider/src/providers/blockchain_provider.rs

```diff
@@ -0,0 +1,817 @@
+
+    fn fill_cfg_env_with_header<EvmConfig>(
+        &self,
+        cfg: &mut CfgEnvWithHandlerCfg,
+        header: &Header,
+        evm_config: EvmConfig,
+    ) -> ProviderResult<()>
+    where
+        EvmConfig: ConfigureEvmEnv,
+    {
+        self.database.provider()?.fill_cfg_env_with_header(cfg, header, evm_config)
+    }
+}
+
+impl<DB> PruneCheckpointReader for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    fn get_prune_checkpoint(
+        &self,
+        segment: PruneSegment,
+    ) -> ProviderResult<Option<PruneCheckpoint>> {
+        self.database.provider()?.get_prune_checkpoint(segment)
+    }
+
+    fn get_prune_checkpoints(&self) -> ProviderResult<Vec<(PruneSegment, PruneCheckpoint)>> {
+        self.database.provider()?.get_prune_checkpoints()
+    }
+}
+
+impl<DB> ChainSpecProvider for BlockchainProvider2<DB>
+where
+    DB: Send + Sync,
+{
+    fn chain_spec(&self) -> Arc<ChainSpec> {
+        self.database.chain_spec()
+    }
+}
+
+impl<DB> StateProviderFactory for BlockchainProvider2<DB>
+where
+    DB: Database,
+{
+    /// Storage provider for latest block
+    fn latest(&self) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", "Getting latest block state provider");
+        self.database.latest()
+    }
+
+    fn history_by_block_number(
+        &self,
+        block_number: BlockNumber,
+    ) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", ?block_number, "Getting history by block number");
+        self.ensure_canonical_block(block_number)?;
+        self.database.history_by_block_number(block_number)
+    }
+
+    fn history_by_block_hash(&self, block_hash: BlockHash) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", ?block_hash, "Getting history by block hash");
+        self.database.history_by_block_hash(block_hash)
+    }
+
+    fn state_by_block_hash(&self, block: BlockHash) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", ?block, "Getting state by block hash");
+        let mut state = self.history_by_block_hash(block);
+
+        // we failed to get the state by hash, from disk, hash block be the pending block
+        if state.is_err() {
+            if let Ok(Some(pending)) = self.pending_state_by_hash(block) {
+                // we found pending block by hash
+                state = Ok(pending)
+            }
+        }
+
+        state
+    }
+
+    /// Returns the state provider for pending state.
+    ///
+    /// If there's no pending block available then the latest state provider is returned:
+    /// [`Self::latest`]
+    fn pending(&self) -> ProviderResult<StateProviderBox> {
+        trace!(target: "providers::blockchain", "Getting provider for pending state");
+
+        if let Some(block) = self.canonical_in_memory_state.pending_block_num_hash() {
+            if let Ok(pending) = self.tree.pending_state_provider(block.hash) {
```

Comment:
Not sure what to do with this `pending_state_provider`, should we implement it in the in memory state?

Reply:
I think we'd need to check if there's a pending block in the canonical memory state, if not this should fallback to latest

Reply:
this is similar to the question before, this method returns a `ProviderResult<StateProviderBox>` how do we obtain a `StateProvider` including in memory information? I think this affects all the methods in the `StateProviderFactory` impl